### ISSUE TITLE
Cannot print DataFrame with dtype `List(Struct(Categorical))`

### ISSUE THREAD

**Issue by stinodego** [pola-rs/polars#12122]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df = pl.DataFrame(
    {"struct": [[{"some_column": "a"}]]},
    schema={"struct": pl.List(pl.Struct([pl.Field("some_column", pl.Categorical)]))},
)
print(df)
```

### Log output

```shell
thread '<unnamed>' panicked at /home/stijn/code/polars/crates/polars-core/src/chunked_array/ops/any_value.rs:56:22:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("cannot construct Categorical from these categories; at least one of them is out of bounds"))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/home/stijn/code/polars/py-polars/repro.py", line 9, in <module>
    print(df)
  File "/home/stijn/code/polars/py-polars/polars/dataframe/frame.py", line 1511, in __str__
    return self._df.as_str()
           ^^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("cannot construct Categorical from these categories; at least one of them is out of bounds"))
```

### Issue description

See example.

### Expected behavior

Example should run.

### Installed versions

main branch

**Comment by cmdlineluser** [comment]
There appears to be some issue in general: #10046

With a StringCache there are no errors:
```python
pl.enable_string_cache(True)

df = pl.DataFrame(
    {"struct": [[{"some_column": "a"}]]},
    schema={"struct": pl.List(pl.Struct([pl.Field("some_column", pl.Categorical)]))},
)

print(df)

# shape: (1, 1)
# ┌─────────────────┐
# │ struct          │
# │ ---             │
# │ list[struct[1]] │
# ╞═════════════════╡
# │ [{"a"}]         │
# └─────────────────┘
```

**Comment by stinodego** [comment]
> There appears to be some issue in general: https://github.com/pola-rs/polars/issues/10046

That's a different issue, but indeed also related to nested categoricals.

**Comment by ritchie46** [comment]
@stinodego do you know how to reproduce this in Rust? I am not sure where we enter the Rust side. As we somehow create invaldid dtypes.

**Comment by stinodego** [comment]
Looks like the problem has 'moved'. The code in the issue description now gives a separate error. Looks like there are two issues actually.

Here is a simpler reproducer that shows casting nested categoricals goes wrong:

```python
import polars as pl

data = [[{"some_column": "a"}]]
dtype = pl.List(pl.Struct({"some_column": pl.Categorical}))

s = pl.Series("a", data, dtype=dtype)  # This works fine
s.cast(dtype)
```
```
thread '<unnamed>' panicked at crates/polars-core/src/series/from.rs:332:25:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/home/stijn/code/polars/py-polars/repro.py", line 7, in <module>
    s.cast(dtype)
  File "/home/stijn/code/polars/py-polars/polars/series/series.py", line 4109, in cast
    return self._from_pyseries(self._s.cast(dtype, strict))
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: called `Option::unwrap()` on a `None` value
```

And a separate issue (the one originally causing this issue) is printing a Series with this type:
```
import polars as pl

data = [[{"some_column": "a"}]]
dtype = pl.List(pl.Struct({"some_column": pl.Categorical}))

s = pl.Series("a", data, dtype=dtype)  # Still works fine
print(s)
```
```
thread '<unnamed>' panicked at crates/polars-core/src/chunked_array/ops/any_value.rs:56:22:
called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("cannot construct Categorical from these categories; at least one of them is out of bounds"))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "/home/stijn/code/polars/py-polars/repro.py", line 7, in <module>
    print(s)
  File "/home/stijn/code/polars/py-polars/polars/series/series.py", line 629, in __str__
    s_repr: str = self._s.as_str()
                  ^^^^^^^^^^^^^^^^
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("cannot construct Categorical from these categories; at least one of them is out of bounds"))
```

**Comment by Object905** [comment]
Seems to be fixed on 1.7.1
